platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileCompactor.java

```diff
@@ -0,0 +1,479 @@
+
+    /**
+     * A lock used for synchronization between snapshots and compactions. While a compaction is in
+     * progress, it runs on its own without any synchronization. However, a few critical sections
+     * are protected with this lock: to create a new compaction writer/reader when compaction is
+     * started, to copy data items to the current writer and update the corresponding index item,
+     * and to close the compaction writer. This mechanism allows snapshots to effectively put
+     * compaction on hold, which is critical as snapshots should be as fast as possible, while
+     * compactions are just background processes.
+     */
+    private final Semaphore snapshotCompactionLock = new Semaphore(1);
+
+    /** Start time of the current compaction, or null if compaction isn't running */
+    private final AtomicReference<Instant> currentCompactionStartTime = new AtomicReference<>();
+    /**
+     * Current data file writer during compaction, or null if compaction isn't running. The writer
+     * is created at compaction start. If compaction is interrupted by a snapshot, the writer is
+     * closed before the snapshot, and then a new writer / new file is created after the snapshot is
+     * taken.
+     */
+    private final AtomicReference<DataFileWriter<?>> currentCompactionWriter = new AtomicReference<>();
+    /** Currrent data file reader for the compaction writer above. */
+    private final AtomicReference<DataFileReader<?>> currentCompactionReader = new AtomicReference<>();
+    /**
+     * The list of new files created during compaction. Usually, all files to process are compacted
+     * to a single new file, but if compaction is interrupted by a snapshot, there may be more than
+     * one file created.
+     */
+    private final List<Path> newCompactedFiles = new ArrayList<>();
+
+    /**
+     * Indicates whether compaction is in progress at the time when {@link #pauseCompaction()}
+     * is called. This flag is then checked in {@link DataFileCompactor#resumeCompaction()} )} to start a new
+     * compacted file or not.
+     */
+    private final AtomicBoolean compactionWasInProgress = new AtomicBoolean(false);
+
+    /**
+     * This variable keeps track of the compaction level that was in progress at the time when it was suspended.
+     * Once the compaction is resumed, this level is used to start a new compacted file, and then it's reset to 0.
+     */
+    private final AtomicInteger compactionLevelInProgress = new AtomicInteger(0);
+
+    /** When was the last medium-sized merge, only touched from single merge thread. */
+    private Instant lastMediumMerge;
+
+    /** When was the last full merge, only touched from single merge thread. */
+    private Instant lastFullMerge;
+
+    /** A nanosecond-precise Clock */
+    private final NanoClock clock = new NanoClock();
+
+    /** The data file collection to compact */
+    private final DataFileCollection<?> dataFileCollection;
```

Comment:
This is the most important field in the class. Important fields are usually placed to the top